chapters/part2/hyperparameter-optimization.md

```diff
@@ -0,0 +1,470 @@
+In each trial step in optimization, the objective function is called and does the following steps:
+
+1. Train a model (`executor.run()`)
+2. Return a target metric on validation data (`executor.run()` returns the specified metric)
+
+Once we've written an objective function, we can write a script for launching optimization.
+In Optuna, we create a study object and pass the objective function to the `optimize()` method as follows.
+You can specify something a number of parameters for the hyperparameter optimization process:
+- a way to save a result of optimization
+- a sampler for searching hyperparameters (`TPESampler` is based on Bayesian Optimization)
+- direction for optimizing (maximize or minimize)
+- number of jobs for distributed training
+- timeout
+and more.  An example launch script is below.
+
+```python
+if __name__ == '__main__':
+    study = optuna.create_study(
+        storage="sqlite:///result/trial.db",  # save results in DB
+        sampler=optuna.samplers.TPESampler(seed=24),
+        study_name="optuna_allennlp",
+        direction="maximize",
+    )
+
+    timeout = 60 * 60 * 10  # timeout (sec): 60*60*10 sec => 10 hours
+    study.optimize(
+        objective,
+        n_jobs=1,  # number of processes in parallel execution
+        n_trials=30,  # number of trials to train a model
+        timeout=timeout,  # threshold for executing time (sec)
+    )
+```
+
+</exercise>
+
+
+<exercise id="6" title="Pruning poor performing trials">
+
+Hyperparameter optimization often takes a long time to find good hyperparameters.
+If you can find and stop unpromising trials with bad hyperparameters, you can reduce the time of optimization and get good hyperparameters faster.
+Stopping unpromising trials is so-called `pruning`.
+The following illustration shows an example of pruning.
+Although the final curves at the left side of the picture are not available before finishing hyperparameter optimization,
+a pruner evaluates at each epoch how promising each trial will be and stops if it predicts to be unpromising.
+
+<img src="/part2/hyperparameter-optimization-with-optuna/illustration_of_pruning.jpg" alt="Illustration of Pruning" />
+
+You can use `AllenNLPPruningCallback` that is the new feature of Optuna,
+which allows users to prune unpromising trials with algorithms implemented in Optuna.
+`AllenNLPPruningCallback` is the interface to provide a way to use these pruning algorithms.
+
+You can enable a pruning callback by adding `optuner_pruner` to `epoch_callbacks` in your jsonnet configuration.
+A pruner determines whether it prune a training in each epoch,
+based on the `metrics` specified in initializing `AllenNLPExecutor`.
+
+```json
+   epoch_callbacks: [
+    {
+      type: 'optuna_pruner',
+    },
+  ],
+```
+
+After enabling pruning callback, you have to specify a pruner you want to use.
+In Optuna, some effient algorithms such as [SuccessiveHaving](https://arxiv.org/abs/1502.07943)
```

Comment:
```suggestion
In Optuna, some effient algorithms such as [SuccessiveHalving](https://arxiv.org/abs/1502.07943)
```

Reply:
Oh, I'm really sorry... thank you for catching.